lib/kernel/doc/guides/socket_usage.md

```diff
@@ -127,6 +127,262 @@ And finally, its possible to override the global default when creating a socket
+		{ok, Data} ->
+		    echo(Sock, Data),
+		    handler_loop(Sock, ?DEFAULT_STATE);
+
+		{select, NewSelectInfo} ->
+		    warning_msg("select from select: "
+			        "~n   Socket Info: ~p"
+			        "~n", [(catch socket:info(Sock))]),
+		    handler_loop(Sock, NewSelectInfo);
+
+		{error, Reason} ->
+		    error_msg("Failed receive data (after select message): "
+			      "~n   Reason:      ~p"
+			      "~n   Socket Info: ~p"
+			      "~n", [Reason, (catch socket:info(Sock))]),
+		    erlang:error({recv_failed, Reason})
+	    end
+    end;
+handler_loop(Sock, ?COMPLETION_INFO(CompletionHandle)) ->
+    receive
+	?COMPLETION_MSG(Sock, CompletionHandle, CompletionStatus) ->
+	    case CompletionStatus of
+		{ok, Data} ->
+		    echo(Sock, Data),
+		    handler_loop(Sock, ?DEFAULT_STATE);
+		{error, Reason} ->
+		    error_msg("Failed receive (completion) data: "
+			      "~n   Reason:      ~p"
+			      "~n   Socket Info: ~p"
+			      "~n", [Reason, (catch socket:info(Sock))]),
+		    erlang:error({recv_failed, Reason})
+	    end
+    end.
+
+echo(Sock, Data) when is_binary(Data) ->
+    ok = socket:send(Sock, Data, infinity),
+    echo_msg("~s", [binary_to_list(Data)]).
+
+```
+
+### The client:
+
+
+```erlang
+-export([client/2, client/3]).
+
+client(SPort, Msg)
+  when is_integer(SPort) andalso (SPort > 0) ->
+    client(SPort, get_local_addr(inet), Msg).
+
+-spec client(SPort, SAddr, Msg) -> ok when
+      SPort :: inet:port_number(),
+      SAddr :: inet:ip_address(),
+      Msg   :: string() | binary().
+
+client(SPort, SAddr, Msg)
+  when is_integer(SPort) andalso (SPort > 0) andalso
+       is_tuple(SAddr) andalso
+       (is_list(Msg) orelse is_binary(Msg)) ->
+    Family     = which_family(SAddr),
+    {ok, Sock} = socket:open(Family, stream, tcp),
+    SSockAddr  = #{family => Family,
+		   addr   => SAddr,
+		   port   => SPort},
+    ok = socket:connect(Sock, SSockAddr),
+    client_echo(Sock, Msg).
+
+client_echo(Sock, Msg) when is_list(Msg) ->
+    client_echo(Sock, list_to_binary(Msg));
+client_echo(Sock, Msg) when is_binary(Msg) ->
+    ok = socket:send(Sock, Msg, infinity),
+    {ok, Msg} = socket:recv(Sock, byte_size(Msg), infinity),
+    ok.
+
+```
+
+### Some common functions:
+
+```erlang
+echo_msg(F, A) ->
+    msg("ECHO", F, A).
+
+warning_msg(F, A) ->
+    msg("WARNING", F, A).
+
+error_msg(F, A) ->
+  msg("ERROR", F, A).
+
+msg(Header, F, A) ->    
+    io:format("** ~s **"
+	      "~n" ++ F ++ "~n", [Header|A]).
+
+get_local_addr(Family) ->
+    Filter =
+	fun(#{addr  := #{family := Fam},
+	      flags := Flags}) ->
+		(Fam =:= Family) andalso (not lists:member(loopback, Flags));
+	   (_) ->
+		false
+	end,
+    {ok, [SockAddr|_]} = net:getifaddrs(Filter),
+    #{addr := #{addr := Addr}} = SockAddr,
+    Addr.
+
+which_family(Addr) when is_tuple(Addr) andalso (tuple_size(Addr) =:= 4) ->
+    inet;
+which_family(Addr) when is_tuple(Addr) andalso (tuple_size(Addr) =:= 8) ->
+    inet6.
```

Comment:
Maybe make use of `inet:is_ipv4_address/1` and `inet:is_ipv6_address/1` instead?

Otherwise, if you want to keep the above, the `is_tuple` part of the guard is superfluous since `tuple_size` will fail if the argument is not a tuple.